# Data Preparation

## Import Libraries

In [1]:
import numpy as np
import pandas as pd


## Import Data

In [2]:
loan_data_backup = pd.read_csv('/Users/rutujajadhav/Desktop/Credit Risk Udemy/loan_data_2007_2014.csv')

/var/folders/s3/mby87d114_g07sh3cl6ysz2r0000gn/T/ipykernel_1204/348723264.py:1: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.
  loan_data_backup = pd.read_csv('/Users/rutujajadhav/Desktop/Credit Risk Udemy/loan_data_2007_2014.csv')


In [3]:
loan_data = loan_data_backup.copy()

## Explore Data

In [4]:
loan_data

Unnamed: 0       id  member_id  loan_amnt  funded_amnt  \
0                0  1077501    1296599       5000         5000   
1                1  1077430    1314167       2500         2500   
2                2  1077175    1313524       2400         2400   
3                3  1076863    1277178      10000        10000   
4                4  1075358    1311748       3000         3000   
...            ...      ...        ...        ...          ...   
466280      466280  8598660    1440975      18400        18400   
466281      466281  9684700   11536848      22000        22000   
466282      466282  9584776   11436914      20700        20700   
466283      466283  9604874   11457002       2000         2000   
466284      466284  9199665   11061576      10000        10000   

        funded_amnt_inv        term  int_rate  installment grade  ...  \
0                4975.0   36 months     10.65       162.87     B  ...   
1                2500.0   60 months     15.27        59.83     C  ...   
2                2400.0   36 months     15.96        84.33     C  ...   
3               10000.0   36 months     13.49       339.31     C  ...   
4                3000.0   60 months     12.69        67.79     B  ...   
...                 ...         ...       ...          ...   ...  ...   
466280          18400.0   60 months     14.47       432.64     C  ...   
466281          22000.0   60 months     19.97       582.50     D  ...   
466282          20700.0   60 months     16.99       514.34     D  ...   
466283           2000.0   36 months      7.90        62.59     A  ...   
466284           9975.0   36 months     19.20       367.58     D  ...   

       total_bal_il il_util open_rv_12m open_rv_24m  max_bal_bc all_util  \
0               NaN     NaN         NaN         NaN         NaN      NaN   
1               NaN     NaN         NaN         NaN         NaN      NaN   
2               NaN     NaN         NaN         NaN         NaN      NaN   
3               NaN     NaN         NaN         NaN         NaN      NaN   
4               NaN     NaN         NaN         NaN         NaN      NaN   
...             ...     ...         ...         ...         ...      ...   
466280          NaN     NaN         NaN         NaN         NaN      NaN   
466281          NaN     NaN         NaN         NaN         NaN      NaN   
466282          NaN     NaN         NaN         NaN         NaN      NaN   
466283          NaN     NaN         NaN         NaN         NaN      NaN   
466284          NaN     NaN         NaN         NaN         NaN      NaN   

       total_rev_hi_lim inq_fi total_cu_tl inq_last_12m  
0                   NaN    NaN         NaN          NaN  
1                   NaN    NaN         NaN          NaN  
2                   NaN    NaN         NaN          NaN  
3                   NaN    NaN         NaN          NaN  
4                   NaN    NaN         NaN          NaN  
...                 ...    ...         ...          ...  
466280          29900.0    NaN         NaN          NaN  
466281          39400.0    NaN         NaN          NaN  
466282          13100.0    NaN         NaN          NaN  
466283          53100.0    NaN         NaN          NaN  
466284          16000.0    NaN         NaN          NaN  

[466285 rows x 75 columns]

In [5]:
loan_data.head()

Unnamed: 0       id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0           0  1077501    1296599       5000         5000           4975.0   
1           1  1077430    1314167       2500         2500           2500.0   
2           2  1077175    1313524       2400         2400           2400.0   
3           3  1076863    1277178      10000        10000          10000.0   
4           4  1075358    1311748       3000         3000           3000.0   

         term  int_rate  installment grade  ... total_bal_il il_util  \
0   36 months     10.65       162.87     B  ...          NaN     NaN   
1   60 months     15.27        59.83     C  ...          NaN     NaN   
2   36 months     15.96        84.33     C  ...          NaN     NaN   
3   36 months     13.49       339.31     C  ...          NaN     NaN   
4   60 months     12.69        67.79     B  ...          NaN     NaN   

  open_rv_12m open_rv_24m  max_bal_bc all_util total_rev_hi_lim inq_fi  \
0         NaN         NaN         NaN      NaN              NaN    NaN   
1         NaN         NaN         NaN      NaN              NaN    NaN   
2         NaN         NaN         NaN      NaN              NaN    NaN   
3         NaN         NaN         NaN      NaN              NaN    NaN   
4         NaN         NaN         NaN      NaN              NaN    NaN   

  total_cu_tl inq_last_12m  
0         NaN          NaN  
1         NaN          NaN  
2         NaN          NaN  
3         NaN          NaN  
4         NaN          NaN  

[5 rows x 75 columns]

In [6]:
loan_data.tail()

Unnamed: 0       id  member_id  loan_amnt  funded_amnt  \
466280      466280  8598660    1440975      18400        18400   
466281      466281  9684700   11536848      22000        22000   
466282      466282  9584776   11436914      20700        20700   
466283      466283  9604874   11457002       2000         2000   
466284      466284  9199665   11061576      10000        10000   

        funded_amnt_inv        term  int_rate  installment grade  ...  \
466280          18400.0   60 months     14.47       432.64     C  ...   
466281          22000.0   60 months     19.97       582.50     D  ...   
466282          20700.0   60 months     16.99       514.34     D  ...   
466283           2000.0   36 months      7.90        62.59     A  ...   
466284           9975.0   36 months     19.20       367.58     D  ...   

       total_bal_il il_util open_rv_12m open_rv_24m  max_bal_bc all_util  \
466280          NaN     NaN         NaN         NaN         NaN      NaN   
466281          NaN     NaN         NaN         NaN         NaN      NaN   
466282          NaN     NaN         NaN         NaN         NaN      NaN   
466283          NaN     NaN         NaN         NaN         NaN      NaN   
466284          NaN     NaN         NaN         NaN         NaN      NaN   

       total_rev_hi_lim inq_fi total_cu_tl inq_last_12m  
466280          29900.0    NaN         NaN          NaN  
466281          39400.0    NaN         NaN          NaN  
466282          13100.0    NaN         NaN          NaN  
466283          53100.0    NaN         NaN          NaN  
466284          16000.0    NaN         NaN          NaN  

[5 rows x 75 columns]

In [7]:
loan_data.columns.values

array(['Unnamed: 0', 'id', 'member_id', 'loan_amnt', 'funded_amnt',
       'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade',
       'sub_grade', 'emp_title', 'emp_length', 'home_ownership',
       'annual_inc', 'verification_status', 'issue_d', 'loan_status',
       'pymnt_plan', 'url', 'desc', 'purpose', 'title', 'zip_code',
       'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line',
       'inq_last_6mths', 'mths_since_last_delinq',
       'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal',
       'revol_util', 'total_acc', 'initial_list_status', 'out_prncp',
       'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv',
       'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee',
       'recoveries', 'collection_recovery_fee', 'last_pymnt_d',
       'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d',
       'collections_12_mths_ex_med', 'mths_since_last_major_derog',
       'policy_code', 'application_type', 'annual_inc_joint', 'dti_joint',
       'veri

In [8]:
loan_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 466285 entries, 0 to 466284
Data columns (total 75 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Unnamed: 0                   466285 non-null  int64  
 1   id                           466285 non-null  int64  
 2   member_id                    466285 non-null  int64  
 3   loan_amnt                    466285 non-null  int64  
 4   funded_amnt                  466285 non-null  int64  
 5   funded_amnt_inv              466285 non-null  float64
 6   term                         466285 non-null  object 
 7   int_rate                     466285 non-null  float64
 8   installment                  466285 non-null  float64
 9   grade                        466285 non-null  object 
 10  sub_grade                    466285 non-null  object 
 11  emp_title                    438697 non-null  object 
 12  emp_length                   445277 non-null  object 
 13 

Preprocessing for continuous variables

In [9]:
loan_data['emp_length'].value_counts()

emp_length
10+ years    150049
2 years       41373
3 years       36596
< 1 year      36265
5 years       30774
1 year        29622
4 years       28023
7 years       26180
6 years       26112
8 years       22395
9 years       17888
Name: count, dtype: int64

In [10]:
loan_data['emp_length_int']=loan_data['emp_length'].str.replace('+ years',' ')
loan_data['emp_length_int'] = loan_data['emp_length_int'].str.replace('< 1 year','0')
loan_data['emp_length_int'] = loan_data['emp_length_int'].str.replace(' years',' ')
loan_data['emp_length_int'] = loan_data['emp_length_int'].str.replace(' year',' ')
loan_data['emp_length_int'] = loan_data['emp_length_int'].fillna(0)

In [11]:
print(loan_data.dtypes['emp_length_int'])

object


In [12]:
loan_data.loc[0:6, ['emp_length','emp_length_int']]

emp_length emp_length_int
0  10+ years            10 
1   < 1 year              0
2  10+ years            10 
3  10+ years            10 
4     1 year             1 
5    3 years             3 
6    8 years             8

In [13]:
loan_data['emp_length_int']= pd.to_numeric(loan_data['emp_length_int'])

In [14]:
loan_data.dtypes['emp_length_int']

dtype('int64')

Term variable pre processing

In [15]:
loan_data['term'].value_counts()

term
36 months    337953
60 months    128332
Name: count, dtype: int64

In [16]:
loan_data['term_int'] = loan_data['term'].str.replace(' months',' ')

In [17]:
loan_data[['term_int','term']].head()

term_int        term
0      36    36 months
1      60    60 months
2      36    36 months
3      36    36 months
4      60    60 months

In [18]:
loan_data.dtypes['term_int']

dtype('O')

In [19]:
loan_data['term_int']=pd.to_numeric(loan_data['term_int'])
loan_data.dtypes['term_int']

dtype('int64')

Preprocessing of time fields

In [20]:
loan_data['earliest_cr_line']

0         Jan-85
1         Apr-99
2         Nov-01
3         Feb-96
4         Jan-96
           ...  
466280    Apr-03
466281    Jun-97
466282    Dec-01
466283    Feb-03
466284    Feb-00
Name: earliest_cr_line, Length: 466285, dtype: object

In [21]:
loan_data['earliest_cr_line_datetime']= pd.to_datetime(loan_data['earliest_cr_line'], format='%b-%y')

In [22]:
loan_data['earliest_cr_line_datetime']

0        1985-01-01
1        1999-04-01
2        2001-11-01
3        1996-02-01
4        1996-01-01
            ...    
466280   2003-04-01
466281   1997-06-01
466282   2001-12-01
466283   2003-02-01
466284   2000-02-01
Name: earliest_cr_line_datetime, Length: 466285, dtype: datetime64[ns]

In [23]:
loan_data['duration_since_earliest_cr_line'] = pd.to_datetime('2017-12-01') - loan_data['earliest_cr_line_datetime']

loan_data['duration_since_earliest_cr_line']

0        12022 days
1         6819 days
2         5874 days
3         7974 days
4         8005 days
            ...    
466280    5358 days
466281    7488 days
466282    5844 days
466283    5417 days
466284    6513 days
Name: duration_since_earliest_cr_line, Length: 466285, dtype: timedelta64[ns]

Converting duration since earliest credit line from days to months. For this we are converting the values from days to months, and also the data type from date time to numeric. np.timedelta64(30,'D') divides the days by 30 days to convert to months.

In [24]:
loan_data['months_since_earliest_cr_line'] = loan_data['duration_since_earliest_cr_line']/np.timedelta64(30,'D')

loan_data['months_since_earliest_cr_line'] = round(loan_data['months_since_earliest_cr_line'])

loan_data['months_since_earliest_cr_line']

0         401.0
1         227.0
2         196.0
3         266.0
4         267.0
          ...  
466280    179.0
466281    250.0
466282    195.0
466283    181.0
466284    217.0
Name: months_since_earliest_cr_line, Length: 466285, dtype: float64

In [25]:
loan_data['months_since_earliest_cr_line'].describe()

count    466256.000000
mean        242.979220
std          95.344015
min        -621.000000
25%         186.000000
50%         228.000000
75%         289.000000
max         596.000000
Name: months_since_earliest_cr_line, dtype: float64

The min value for months since earliest credit line is negative. which means past date was not accurately parsed which we initially convert to date time. some of these years must have been converted to future. Investigate which rows have negative months since earliest credit line and observe the parsed date time value for these.

We can see the years corresponding to negative months since earliest start date are in future - 2062, etc.

In [26]:
loan_data[['earliest_cr_line','earliest_cr_line_datetime','months_since_earliest_cr_line']][loan_data['months_since_earliest_cr_line']<0]

earliest_cr_line earliest_cr_line_datetime  \
1580             Sep-62                2062-09-01   
1770             Sep-68                2068-09-01   
2799             Sep-64                2064-09-01   
3282             Sep-67                2067-09-01   
3359             Feb-65                2065-02-01   
...                 ...                       ...   
464003           Jan-68                2068-01-01   
464260           Jul-66                2066-07-01   
465100           Oct-67                2067-10-01   
465500           Sep-67                2067-09-01   
465655           Jan-56                2056-01-01   

        months_since_earliest_cr_line  
1580                           -545.0  
1770                           -618.0  
2799                           -569.0  
3282                           -606.0  
3359                           -574.0  
...                               ...  
464003                         -610.0  
464260                         -591.0  
465100                         -607.0  
465500                         -606.0  
465655                         -464.0  

[1169 rows x 3 columns]

adjusting for incorrectly parsed future dates: either go back to date time parse step and replace these years to reflect 1960s instead of 2060s or delete all the 1169 rows where the months since cr line are negative. OR.. manipulate these rows to change the negative difference to another value. 

Since these dates belong to the 1960s (which is furthest away from our current date of 1 Dec 2017, we can substitute the negative months, presumably from 1960s in the months_since_earliest_cr column for the largest difference in months available. larger the difference in months => closer is the date to 1960s.

In [27]:
loan_data.loc[loan_data['months_since_earliest_cr_line']<0,'months_since_earliest_cr_line'] = loan_data['months_since_earliest_cr_line'].max()

loan_data['months_since_earliest_cr_line'].describe()

count    466256.000000
mean        245.925734
std          87.722725
min          74.000000
25%         187.000000
50%         229.000000
75%         289.000000
max         596.000000
Name: months_since_earliest_cr_line, dtype: float64

preprocessing issue_date

In [28]:
loan_data.loc[:6,'issue_d']

0    Dec-11
1    Dec-11
2    Dec-11
3    Dec-11
4    Dec-11
5    Dec-11
6    Dec-11
Name: issue_d, dtype: object

In [29]:
loan_data['issue_d_datetime'] = pd.to_datetime(loan_data['issue_d'], format='%b-%y')
loan_data.dtypes['issue_d_datetime']


dtype('<M8[ns]')

In [30]:
loan_data['days_since_issue_date'] = pd.to_datetime('2017-12-01')-loan_data['issue_d_datetime']
loan_data['days_since_issue_date']

0        2192 days
1        2192 days
2        2192 days
3        2192 days
4        2192 days
            ...   
466280   1430 days
466281   1430 days
466282   1430 days
466283   1430 days
466284   1430 days
Name: days_since_issue_date, Length: 466285, dtype: timedelta64[ns]

In [31]:
loan_data['months_since_issue_date'] = loan_data['days_since_issue_date']/np.timedelta64(30,'D')
loan_data['months_since_issue_date']= round(loan_data['months_since_issue_date'])
loan_data['months_since_issue_date']

0         73.0
1         73.0
2         73.0
3         73.0
4         73.0
          ... 
466280    48.0
466281    48.0
466282    48.0
466283    48.0
466284    48.0
Name: months_since_issue_date, Length: 466285, dtype: float64

In [32]:
loan_data['months_since_issue_date'].describe()

count    466285.000000
mean         52.265887
std          14.385250
min          37.000000
25%          42.000000
50%          48.000000
75%          58.000000
max         128.000000
Name: months_since_issue_date, dtype: float64

PREPROCESSING of Discrete Variables

In [33]:
trial_dummy_categorical = pd.get_dummies(loan_data['grade'])
trial_dummy_categorical

A      B      C      D      E      F      G
0       False   True  False  False  False  False  False
1       False  False   True  False  False  False  False
2       False  False   True  False  False  False  False
3       False  False   True  False  False  False  False
4       False   True  False  False  False  False  False
...       ...    ...    ...    ...    ...    ...    ...
466280  False  False   True  False  False  False  False
466281  False  False  False   True  False  False  False
466282  False  False  False   True  False  False  False
466283   True  False  False  False  False  False  False
466284  False  False  False   True  False  False  False

[466285 rows x 7 columns]

The get_dummies, extracts categories from the column and makes another the table. The col represents the categories with True value for the row where its True - all other categories will be False. Ex Row 1, Grade is C.

To preprocess the other categorical variables, we will generate such dataframes for each variable qnd store them in a list. Then we will use concat to join the list of dataframes into one big dataframe and join this with the original dataframe. 

list_of_dfs= [df1, df2, df3]
df_concated = pd.concat(list_of_dfs, axis = 1) {converting list of dataframes into one dataframe - join this with loan_data. axis = 0 will concat the tables vertically and axis=1 will concat them horizontally}


In [34]:
loan_data_dummies_listofdfs = [pd.get_dummies(loan_data['grade'], prefix='Grade', prefix_sep = ':'),pd.get_dummies(loan_data['sub_grade'], prefix='Sub-Grade', prefix_sep = ':'), pd.get_dummies(loan_data['home_ownership'],prefix='Home_Own', prefix_sep = ':'), pd.get_dummies(loan_data['verification_status'],prefix='Verification_Status', prefix_sep = ':'), pd.get_dummies(loan_data['loan_status'], prefix='Loan_Status', prefix_sep = ':'), pd.get_dummies(loan_data['purpose'],prefix='Purpose', prefix_sep = ''), pd.get_dummies(loan_data['addr_state'], prefix='Addr_State', prefix_sep = ':'), pd.get_dummies(loan_data['initial_list_status'], prefix='initial_list_status', prefix_sep = ':')]


In [35]:
loan_data_dummies_concatenated = pd.concat(loan_data_dummies_listofdfs, axis=1) #axis = 0 will concat the tables vertically and axis=1 will concat them horizontally

In [36]:
type(loan_data_dummies_concatenated)

pandas.core.frame.DataFrame

In [37]:
loan_data = pd.concat([loan_data, loan_data_dummies_concatenated],axis=1 )

In [38]:
loan_data.columns.values

array(['Unnamed: 0', 'id', 'member_id', 'loan_amnt', 'funded_amnt',
       'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade',
       'sub_grade', 'emp_title', 'emp_length', 'home_ownership',
       'annual_inc', 'verification_status', 'issue_d', 'loan_status',
       'pymnt_plan', 'url', 'desc', 'purpose', 'title', 'zip_code',
       'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line',
       'inq_last_6mths', 'mths_since_last_delinq',
       'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal',
       'revol_util', 'total_acc', 'initial_list_status', 'out_prncp',
       'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv',
       'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee',
       'recoveries', 'collection_recovery_fee', 'last_pymnt_d',
       'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d',
       'collections_12_mths_ex_med', 'mths_since_last_major_derog',
       'policy_code', 'application_type', 'annual_inc_joint', 'dti_joint',
       'veri

Managing missing values

In [39]:
pd.options.display.max_rows = None
loan_data.isnull().sum()

Unnamed: 0                                                              0
id                                                                      0
member_id                                                               0
loan_amnt                                                               0
funded_amnt                                                             0
funded_amnt_inv                                                         0
term                                                                    0
int_rate                                                                0
installment                                                             0
grade                                                                   0
sub_grade                                                               0
emp_title                                                           27588
emp_length                                                          21008
home_ownership                        

In [40]:
loan_data['total_rev_hi_lim'].fillna(loan_data['funded_amnt'], inplace=True)
loan_data['total_rev_hi_lim'].isnull().sum()

/var/folders/s3/mby87d114_g07sh3cl6ysz2r0000gn/T/ipykernel_1204/4084969148.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  loan_data['total_rev_hi_lim'].fillna(loan_data['funded_amnt'], inplace=True)


0

In [41]:
loan_data['annual_inc'].isnull().sum()

4

In [42]:
loan_data['annual_inc'].fillna(loan_data['annual_inc'].mean(), inplace=True)

/var/folders/s3/mby87d114_g07sh3cl6ysz2r0000gn/T/ipykernel_1204/4036234326.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  loan_data['annual_inc'].fillna(loan_data['annual_inc'].mean(), inplace=True)


In [43]:
loan_data['annual_inc'].isnull().sum()

0

In [44]:
loan_data['months_since_earliest_cr_line'].isnull().sum()

29

In [45]:
loan_data['months_since_earliest_cr_line'].fillna(0, inplace=True)
loan_data['acc_now_delinq'].fillna(0, inplace=True)
loan_data['total_acc'].fillna(0, inplace=True)
loan_data['pub_rec'].fillna(0, inplace=True)
loan_data['open_acc'].fillna(0, inplace=True)
loan_data['inq_last_6mths'].fillna(0, inplace=True)
loan_data['delinq_2yrs'].fillna(0, inplace=True)
loan_data['emp_length_int'].fillna(0, inplace=True)

/var/folders/s3/mby87d114_g07sh3cl6ysz2r0000gn/T/ipykernel_1204/958730159.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  loan_data['months_since_earliest_cr_line'].fillna(0, inplace=True)
/var/folders/s3/mby87d114_g07sh3cl6ysz2r0000gn/T/ipykernel_1204/958730159.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setti

Designing the PD Model

Assigning the Non default (1- good) and Default (0 - bad) indicators to the loan status. (Create a new column good_bad to store these. select records from loan_Status col which was Charged off, Default type of values and replace them with 0, and others with 1.

In [46]:
loan_data['loan_status'].value_counts(normalize=True, ascending=False)

loan_status
Current                                                0.480878
Fully Paid                                             0.396193
Charged Off                                            0.091092
Late (31-120 days)                                     0.014798
In Grace Period                                        0.006747
Does not meet the credit policy. Status:Fully Paid     0.004263
Late (16-30 days)                                      0.002612
Default                                                0.001784
Does not meet the credit policy. Status:Charged Off    0.001632
Name: proportion, dtype: float64

In [47]:
loan_data['good_bad'] = np.where(loan_data['loan_status'].isin(['Charged Off','Default','Does not meet the credit policy. Status: Charged Off', 'Late (31-120 days)']), 0,1)

In [48]:
loan_data[['loan_status','good_bad']].head()

loan_status  good_bad
0   Fully Paid         1
1  Charged Off         0
2   Fully Paid         1
3   Fully Paid         1
4      Current         1

Splitting the data into training and testing

In [49]:
from sklearn.model_selection import train_test_split


In [50]:
#creating the training (80%) and testing (20%) datasets. There will be 4 data sets in total - training set with predictors (inputs or indp vars), testing set with predictors, training set with target variable, and test set with target variable.

#the train test split method (), just needs : 
# 1. the df selection containing the training columns (we took all cols expect the good_bad col), 2. the df column containing the predictor vars, and 3. fraction of split. This method also shuffles the data randomly between test and train data. But shuffling at every run will tweak the results. So we assign a random seed, so that observations are always shuffled in the same random way.

loan_data_inputs_train, loan_data_inputs_test, loan_data_targets_train, loan_data_targets_test = train_test_split(loan_data.drop('good_bad', axis=1),loan_data['good_bad'], test_size= 0.2, random_state = 42)

In [51]:
loan_data_inputs_train.shape

(373028, 209)

In [52]:
loan_data_inputs_test.shape

(93257, 209)

In [53]:
loan_data_targets_train.shape

(373028,)

In [54]:
loan_data_targets_test.shape

(93257,)

In [55]:
df_inputs_prep = loan_data_inputs_train.copy()
df_targets_prep = loan_data_targets_train.copy()
df1 = pd.concat([df_inputs_prep['grade'], df_targets_prep], axis=1)
df1.head()

grade  good_bad
427211     A         1
206088     C         1
136020     A         1
412305     D         0
36159      C         0

In [56]:
df_category = df1.groupby(df1.columns.values[0],as_index=False)[df1.columns.values[1]].count()

df_category

grade  good_bad
0     A     59759
1     B    109730
2     C    100245
3     D     61498
4     E     28612
5     F     10530
6     G      2654

Calculate WOE = ln[(no good borrowers in a cat/total no good borrowers)/(no bad borrowers in a cat/total no bad borrowers)]

no of good borrowers = prop of good borrowers in a cat * tot no of records in that cat. We already have tot no records in the category as Count. For example in A the 59759 have some good and some bad borrowers. The good borrowers are marked by 1. so the number of good borrowers can be calculated as either the sum of values in the good_bad col for each cat, or by multiplying, mean of good borrowers in the cat * tot number of borrowers in the cat)


In [57]:
df_category = pd.concat([df_category, df1.groupby(df1.columns.values[0], as_index=False)[df1.columns.values[1]].mean()],axis = 1)

df_category

grade  good_bad grade  good_bad
0     A     59759     A  0.961161
1     B    109730     B  0.921589
2     C    100245     C  0.886947
3     D     61498     D  0.848987
4     E     28612     E  0.809940
5     F     10530     F  0.761159
6     G      2654     G  0.748681

In [58]:
df_category = df_category.iloc[:, [0,1,3]]
df_category

grade  good_bad  good_bad
0     A     59759  0.961161
1     B    109730  0.921589
2     C    100245  0.886947
3     D     61498  0.848987
4     E     28612  0.809940
5     F     10530  0.761159
6     G      2654  0.748681

In [59]:
df_category.columns = ['Grade','num_tot_obs','prop_good_in_catg']
df_category

Grade  num_tot_obs  prop_good_in_catg
0     A        59759           0.961161
1     B       109730           0.921589
2     C       100245           0.886947
3     D        61498           0.848987
4     E        28612           0.809940
5     F        10530           0.761159
6     G         2654           0.748681

In [60]:
df_category['num_good_in_catg'] = df_category['prop_good_in_catg']* df_category['num_tot_obs']

df_category['prop_bad_in_catg'] = 1 - df_category['prop_good_in_catg']

df_category['num_bad_in_catg'] = df_category['prop_bad_in_catg']*df_category['num_tot_obs']

df_category['prop_good_wrt_total_good'] = df_category['num_good_in_catg']/df_category['num_good_in_catg'].sum()

df_category['prop_bad_wrt_total_bad'] = df_category['num_bad_in_catg']/df_category['num_bad_in_catg'].sum()

df_category


Grade  num_tot_obs  prop_good_in_catg  num_good_in_catg  prop_bad_in_catg  \
0     A        59759           0.961161           57438.0          0.038839   
1     B       109730           0.921589          101126.0          0.078411   
2     C       100245           0.886947           88912.0          0.113053   
3     D        61498           0.848987           52211.0          0.151013   
4     E        28612           0.809940           23174.0          0.190060   
5     F        10530           0.761159            8015.0          0.238841   
6     G         2654           0.748681            1987.0          0.251319   

   num_bad_in_catg  prop_good_wrt_total_good  prop_bad_wrt_total_bad  
0           2321.0                  0.172557                0.057787  
1           8604.0                  0.303807                0.214216  
2          11333.0                  0.267113                0.282161  
3           9287.0                  0.156854                0.231221  
4           5438.0                  0.069620                0.135392  
5           2515.0                  0.024079                0.062617  
6            667.0                  0.005969                0.016606

In [61]:
df_category['WOE']= np.log(df_category['prop_good_wrt_total_good']/df_category['prop_bad_wrt_total_bad'])

df_category['WOE']

0    1.093973
1    0.349405
2   -0.054807
3   -0.388057
4   -0.665115
5   -0.955693
6   -1.023144
Name: WOE, dtype: float64

In [62]:
df_category

Grade  num_tot_obs  prop_good_in_catg  num_good_in_catg  prop_bad_in_catg  \
0     A        59759           0.961161           57438.0          0.038839   
1     B       109730           0.921589          101126.0          0.078411   
2     C       100245           0.886947           88912.0          0.113053   
3     D        61498           0.848987           52211.0          0.151013   
4     E        28612           0.809940           23174.0          0.190060   
5     F        10530           0.761159            8015.0          0.238841   
6     G         2654           0.748681            1987.0          0.251319   

   num_bad_in_catg  prop_good_wrt_total_good  prop_bad_wrt_total_bad       WOE  
0           2321.0                  0.172557                0.057787  1.093973  
1           8604.0                  0.303807                0.214216  0.349405  
2          11333.0                  0.267113                0.282161 -0.054807  
3           9287.0                  0.156854                0.231221 -0.388057  
4           5438.0                  0.069620                0.135392 -0.665115  
5           2515.0                  0.024079                0.062617 -0.955693  
6            667.0                  0.005969                0.016606 -1.023144

In [64]:
df_category['IV'] = (df_category['prop_good_wrt_total_good'] - df_category['prop_bad_wrt_total_bad']) * df_category['WOE']

df_category['IV'] = df_category['IV'].sum()

#IV = SUM OF [(prop outcome1 wrt total outcome 1 across all catg - prop outcome2 wrt total outcome 2 across all catg) x WOE]. We first do the product for each row. and then sum them all. therefore IV value is same for all! Its a value for the overall category - Grades, derived form the individual WOEs which we calculated for each class within the overall category.

df_category


Grade  num_tot_obs  prop_good_in_catg  num_good_in_catg  prop_bad_in_catg  \
0     A        59759           0.961161           57438.0          0.038839   
1     B       109730           0.921589          101126.0          0.078411   
2     C       100245           0.886947           88912.0          0.113053   
3     D        61498           0.848987           52211.0          0.151013   
4     E        28612           0.809940           23174.0          0.190060   
5     F        10530           0.761159            8015.0          0.238841   
6     G         2654           0.748681            1987.0          0.251319   

   num_bad_in_catg  prop_good_wrt_total_good  prop_bad_wrt_total_bad  \
0           2321.0                  0.172557                0.057787   
1           8604.0                  0.303807                0.214216   
2          11333.0                  0.267113                0.282161   
3           9287.0                  0.156854                0.231221   
4           5438.0                  0.069620                0.135392   
5           2515.0                  0.024079                0.062617   
6            667.0                  0.005969                0.016606   

        WOE        IV  
0  1.093973  0.278002  
1  0.349405  0.278002  
2 -0.054807  0.278002  
3 -0.388057  0.278002  
4 -0.665115  0.278002  
5 -0.955693  0.278002  
6 -1.023144  0.278002

In [66]:
df_category = df_category.sort_values(['WOE'])
df_category = df_category.reset_index(drop=True)
df_category

Grade  num_tot_obs  prop_good_in_catg  num_good_in_catg  prop_bad_in_catg  \
0     G         2654           0.748681            1987.0          0.251319   
1     F        10530           0.761159            8015.0          0.238841   
2     E        28612           0.809940           23174.0          0.190060   
3     D        61498           0.848987           52211.0          0.151013   
4     C       100245           0.886947           88912.0          0.113053   
5     B       109730           0.921589          101126.0          0.078411   
6     A        59759           0.961161           57438.0          0.038839   

   num_bad_in_catg  prop_good_wrt_total_good  prop_bad_wrt_total_bad  \
0            667.0                  0.005969                0.016606   
1           2515.0                  0.024079                0.062617   
2           5438.0                  0.069620                0.135392   
3           9287.0                  0.156854                0.231221   
4          11333.0                  0.267113                0.282161   
5           8604.0                  0.303807                0.214216   
6           2321.0                  0.172557                0.057787   

        WOE        IV  
0 -1.023144  0.278002  
1 -0.955693  0.278002  
2 -0.665115  0.278002  
3 -0.388057  0.278002  
4 -0.054807  0.278002  
5  0.349405  0.278002  
6  1.093973  0.278002

In [68]:
df_category['diff_prop_good_in_catg'] = df_category['prop_good_in_catg'].diff().abs()

df_category['diff_WOE'] = df_category['WOE'].diff().abs()

df_category

Grade  num_tot_obs  prop_good_in_catg  num_good_in_catg  prop_bad_in_catg  \
0     G         2654           0.748681            1987.0          0.251319   
1     F        10530           0.761159            8015.0          0.238841   
2     E        28612           0.809940           23174.0          0.190060   
3     D        61498           0.848987           52211.0          0.151013   
4     C       100245           0.886947           88912.0          0.113053   
5     B       109730           0.921589          101126.0          0.078411   
6     A        59759           0.961161           57438.0          0.038839   

   num_bad_in_catg  prop_good_wrt_total_good  prop_bad_wrt_total_bad  \
0            667.0                  0.005969                0.016606   
1           2515.0                  0.024079                0.062617   
2           5438.0                  0.069620                0.135392   
3           9287.0                  0.156854                0.231221   
4          11333.0                  0.267113                0.282161   
5           8604.0                  0.303807                0.214216   
6           2321.0                  0.172557                0.057787   

        WOE        IV  diff_prop_good_in_catg  diff_WOE  
0 -1.023144  0.278002                     NaN       NaN  
1 -0.955693  0.278002                0.012477  0.067451  
2 -0.665115  0.278002                0.048781  0.290578  
3 -0.388057  0.278002                0.039047  0.277058  
4 -0.054807  0.278002                0.037960  0.333251  
5  0.349405  0.278002                0.034642  0.404212  
6  1.093973  0.278002                0.039571  0.744568

Function to calculate WOE and IV for any categorical variable.

Input parameters = input data set, col name of the category in the input data set, output data set and col of out put. In our case, the input data set is df_inputs_prep, and output dataset containing just the output column is df_targets_prep.


Steps:
1. gather the parameters - input df, input df col, target df
2. Create a df containing only the required categorical var and target var = overwrite the input df to contain only input df with the desired category column, and concat this with the target df. 
3. Overwrite the df to group by classes in the category. use indexes instead of col names to do this. Set group by as_index to false. first group by class of input wrt count of target, and then concat with group by class of input wrt mean of target. set axis = 1 while using concat.
4. there will be two columns of classes so, now overwrite the df to contain only one col of classes, count, and mean of target.
5. rename count col as num_obs_in_catg, and mean col as prop_good_in_catg (since good is assigned a value of 1, the sum or  mean functions will show values for good only).
5. create a col for num of good in cat = num_obs x prop of good in catg
6.prop bad = 1- prop of good
6. num bad = proo bad x num_obs
7. prop good wrt total good = num good/ sum of num of goods
8. prop bad wrt total bad = num bad/ sum of num of bad
9. WOE = log (prop good wrt total good/ prop bad wrt toal bad)
10. IV = (prop of good wrt tot good - prop of bad wrt total bad)x WOE
11. IV = sum of all IV calculated in prev step
12. sort by WOE in ascending
13. OVERWRITE df to reset_index (otherwise rest index will create a new df) , (drop = True) to drop the current index
14. calculate row-wise diff between prop_good_in catg
15. calculate row-wise diff between WOE

In [69]:
def woe_iv_discrete (df,discrete_var_col,target_df):
    df = pd.concat([df[discrete_var_col],target_df], axis=1)
    
    df= pd.concat([df.groupby(df.columns.values[0],as_index=False)[df.columns.values[1]].count(), df.groupby(df.columns.values[0],as_index=False)[df.columns.values[1]].mean()], axis = 1)
    
    df = df.iloc[:, [0,1,3]]
    
    df.columns = [df.columns.values[0], 'n_tot_obs','propn_good_in_catg']
    
    df['n_good_in_catg'] = df['propn_good_in_catg'] * df['n_tot_obs']
    
    df['propn_bad_in_catg'] = 1 - df['propn_good_in_catg']
    
    df['n_bad_in_catg'] = df['propn_bad_in_catg'] * df['n_tot_obs']
    
    df['propn_good_wrt_tot_good'] = df['n_good_in_catg']/df['n_good_in_catg'].sum()
    
    df['propn_bad_wrt_tot_bad'] = df['n_bad_in_catg']/ df['n_bad_in_catg'].sum()
    
    df['WOE_catg'] = np.log( df['propn_good_wrt_tot_good']/ df['propn_bad_wrt_tot_bad'])
    
    df['IV'] = (df['propn_good_wrt_tot_good'] -  df['propn_bad_wrt_tot_bad'])* df['WOE_catg']
    
    df['IV'] = df['IV'].sum()
    
    df = df.sort_values(['WOE_catg'])
    df = df.reset_index(drop=True)
    
    df['diff_propn_good_in_catg'] = df['propn_good_in_catg'].diff().abs()
    
    df['diff_WOE_catg']= df['WOE_catg'].diff().abs()
    
    return df
    
    
    
    

In [70]:
df_temp = woe_iv_discrete(df_inputs_prep,'grade',df_targets_prep)

df_temp

grade  n_tot_obs  propn_good_in_catg  n_good_in_catg  propn_bad_in_catg  \
0     G       2654            0.748681          1987.0           0.251319   
1     F      10530            0.761159          8015.0           0.238841   
2     E      28612            0.809940         23174.0           0.190060   
3     D      61498            0.848987         52211.0           0.151013   
4     C     100245            0.886947         88912.0           0.113053   
5     B     109730            0.921589        101126.0           0.078411   
6     A      59759            0.961161         57438.0           0.038839   

   n_bad_in_catg  propn_good_wrt_tot_good  propn_bad_wrt_tot_bad  WOE_catg  \
0          667.0                 0.005969               0.016606 -1.023144   
1         2515.0                 0.024079               0.062617 -0.955693   
2         5438.0                 0.069620               0.135392 -0.665115   
3         9287.0                 0.156854               0.231221 -0.388057   
4        11333.0                 0.267113               0.282161 -0.054807   
5         8604.0                 0.303807               0.214216  0.349405   
6         2321.0                 0.172557               0.057787  1.093973   

         IV  diff_propn_good_in_catg  diff_WOE_catg  
0  0.278002                      NaN            NaN  
1  0.278002                 0.012477       0.067451  
2  0.278002                 0.048781       0.290578  
3  0.278002                 0.039047       0.277058  
4  0.278002                 0.037960       0.333251  
5  0.278002                 0.034642       0.404212  
6  0.278002                 0.039571       0.744568